### PR TITLE
Upkeep/code review

### PR DESCRIPTION
**Pull request by rkallos**
This branch focuses on the code review performed on 2017-02-08.

Some of the larger tasks, such as implementing deduplication in C, switching to a trie-like structure for `watch_map`, and using refs instead of atoms for mailbox scanning, are left for the future.

There are a few tasks remaining for this branch:
- [x] Document `create_watch_map`
- [x] Call `load/0` only when starting the application
- [x] Add check for priv directory

### CODE REVIEW

File: c_src/filewatch_inotify.c [hunk 1]
```diff
@@ -9,9 +13,12 @@
...
 #include <erl_driver.h>
 
-#include "macrology.h"
-
-enum { MAX_COOLDOWNS = 12, MSG_LENGTH = 11 };
+enum {
+    MAX_EVENTS = 12,
+    EVENT_TUPLE_LENGTH = 11,
+    PORT_CALL_ARG = 1,
+    WATCH_FLAGS = IN_MOVED_TO | IN_CLOSE_WRITE | IN_ATTRIB
```
Comment (RAttab) [inline c_src/filewatch_inotify.c]: I'm curious, why `IN_ATTRIB`?

File: c_src/filewatch_inotify.c [hunk 2]
```diff
@@ -149,9 +167,55 @@ static ErlDrvSSizeT call(
...
 
+/* Return type to Erlang:
+   {port(), [{watch_descriptor(), filename:basename()}]}
+*/
+static void send_output(ErlDrvData self_)
+{
+    struct instance *self = (struct instance *)self_;
+
+    size_t pos = 0;
+    ErlDrvTermData buf[(MAX_EVENTS * EVENT_TUPLE_LENGTH) + 3];
```
Comment (RAttab) [inline c_src/filewatch_inotify.c]: It's a bit hard to tell what `+3` refers to. A comment or a named constant would be helpful.

File: c_src/filewatch_inotify.c [hunk 3]
```diff
@@ -149,9 +167,55 @@ static ErlDrvSSizeT call(
...
+    struct message *msg;
+    for(uint i = 0; i < self->n_events; ++i) {
+        msg = &(self->msgs[i]);
+        buf[pos++] = ERL_DRV_INT;
+        buf[pos++] = msg->wd;
+        buf[pos++] = ERL_DRV_STRING;
+        buf[pos++] = (ErlDrvTermData) msg->name;
+        buf[pos++] = strlen(msg->name);
+        buf[pos++] = ERL_DRV_TUPLE;
+        buf[pos++] = 2;
```
Comment (RAttab) [inline c_src/filewatch_inotify.c]: What's this `2` again?

File: c_src/filewatch_inotify.c (hunk 1 above)
Comment (rkallos) [reply c_src/filewatch_inotify.c]: IIRC this is a remnant from me trying to debug gateway not updating when the staging bert files did not trigger filewatch when they were rsynced.

File: c_src/filewatch_inotify.c (hunk 3 above)
Comment (rkallos) [reply c_src/filewatch_inotify.c]: The size of the `{filewatch, Term}` tuple.

File: src/filewatch.erl [hunk 4]
```diff
@@ -22,40 +53,38 @@ priv_dir() ->
 
 %% If you put this in on_load, it will immediately get unloaded when that child dies.
 load() ->
-    case erl_ddll:load_driver(priv_dir(), "filewatch") of
+    case erl_ddll:load_driver(priv_dir(), ?DRIVER_NAME) of
```
Comment (RAttab) [inline src/filewatch.erl]: Code changes this line from a string to an atom. Intentional?

File: src/filewatch.erl [hunk 5]
```diff
@@ -22,40 +53,38 @@ priv_dir() ->
...
--spec stop(handle()) -> ok | {error, _}.
-
-stop(Handle) ->
-    Handle ! terminate,
-    ok.
 
 init(Pid, Pairs) ->
     ok = load(),
-    Port = open_port({spawn_driver, "filewatch "}, [in]),
+    Arg = io_lib:format("~p ~p", [?DRIVER_NAME, ?COOLDOWN_MS]),
```
Comment (RAttab) [inline src/filewatch.erl]: Is the driver name required here? Would save you some string parsing in the C code.

Comment (RAttab) [reply src/filewatch.erl]: Since you control the arguments in the erlang code, you can simplify the argument parsing code quite a bit as a result by just assuming certain things (eg. no extra whitespaces).

File: src/filewatch.erl [hunk 6]
```diff
@@ -22,40 +53,38 @@ priv_dir() ->
...
-    watch(#state{pid=Pid, port=Port, map=WatchMap}).
+    [cast(Pid, Term) || {_, Term} <- Pairs],
+    watch(#state{pid = Pid, port = Port, watch_map = WatchMap}).
 
 
 create_watch_map([], _Port, Map) -> Map;
 create_watch_map([{Path, Term} | Rest], Port, Map) ->
+    NewMap = case filelib:is_file(Path) of
+                 true -> add_to_watch_map({Path, Term}, Port, Map);
+                 false -> error_logger:error_msg("~p: Path ~p is not a file",
```
Comment (RAttab) [inline src/filewatch.erl]: Will filewatch fail to watch any files that don't exist when filewatch is started? This behaviour could lead to some  rather confusing situations.

Comment (rkallos) [reply src/filewatch.erl]: Good point. Rather then ensuring that every file exists, it's enough to verify that every parent directory for every file exists, since that's what is actually getting watched by inotify.

File: src/filewatch.erl (hunk 5 above)
Comment (rkallos) [reply src/filewatch.erl]: I'm not quite sure what you mean...

File: src/filewatch.erl (hunk 4 above)
Comment (rkallos) [reply src/filewatch.erl]: `load_driver` accepts `iolist() | atom()`, so it doesn't make much difference.

File: c_src/filewatch_inotify.c [hunk 7]
```diff
@@ -88,27 +105,30 @@ static void stop(ErlDrvData self_)
...
 
     int type;
     int path_len;
-    if (ei_get_type(buf, &index, &type, &path_len) < 0) return -1;
+    if (ei_get_type(buf, &index, &type, &path_len) < 0) goto fail_decode;
 
     assert(path_len >= 0);
-    char path[path_len+1];
+
+    char *path = alloca(path_len+1);
```
Comment (RAttab) [inline c_src/filewatch_inotify.c:127]: Why the change to `alloca`? Was that because the compiler was whining about something? If so, it might be worth indicating it in the code although it doesn't matter much.

File: c_src/filewatch_inotify.c [hunk 8]
```diff
@@ -149,9 +169,55 @@ static ErlDrvSSizeT call(
...
+    buf[pos++] = driver_mk_port(self->port);
+
+    struct message *msg;
+    for(uint i = 0; i < self->n_events; ++i) {
+        msg = &(self->msgs[i]);
+        buf[pos++] = ERL_DRV_INT;
+        buf[pos++] = msg->wd;
+        buf[pos++] = ERL_DRV_STRING;
+        buf[pos++] = (ErlDrvTermData) msg->name;
+        buf[pos++] = strlen(msg->name);
```
Comment (RAttab) [inline c_src/filewatch_inotify.c]: Consider using `strnlen`.

File: c_src/filewatch_inotify.c [hunk 9]
```diff
@@ -149,9 +169,55 @@ static ErlDrvSSizeT call(
...
+    }
+
+    /* Append values designating Erlang type */
+    buf[pos++] = ERL_DRV_NIL;
+    buf[pos++] = ERL_DRV_LIST;
+    buf[pos++] = self->n_events + 1; // +1 for nil
+    buf[pos++] = ERL_DRV_TUPLE;
+    buf[pos++] = BUFFER_TUPLE_SIZE;
+
+    /* Reset self */
```
Comment (RAttab) [inline c_src/filewatch_inotify.c]: This comment doesn't add much and I believe the line itself is pretty clear as to what it does.

File: c_src/macrology.h [hunk 10]
```diff
@@ -1,3 +1,11 @@
+/* o/~ readings in macrology /
+ *     never meant that much to me /
+ *     the language I thought was just so /
+ *     she tells me, we ain't compatible
+ *
+ * Copyright 2015 AdGear Technologies Inc.
```
Comment (RAttab) [inline c_src/macrology.h:6]: :+1:

File: src/filewatch.erl [hunk 11]
```diff
@@ -1,15 +1,49 @@
...
+-export([start/1,
+         start/2,
+         stop/1]).
+
+-type watch_pair() :: {file:name_all(), term()}.
+-export_type([watch_pair/0]).
+
+-define(DRIVER_NAME, filewatch).
+-define(COOLDOWN_MS,
+        case application:get_env(cooldown_ms) of
```
Comment (RAttab) [inline src/filewatch.erl]: Consider `application:get_env/3` which accepts a default parameter as the third argument.

File: src/filewatch.erl (hunk 5 above)
Comment (RAttab) [reply src/filewatch.erl]: Could you just call `io_lib:format("~p", [?COOLDOWN_MS])`. If that's possible, it would make the argument parsing function in the C code simpler.

File: src/filewatch.erl [hunk 12]
```diff
@@ -1,107 +1,134 @@
...
+-type directory() :: file:filename().
 
 -record(state,
         {pid :: pid(),
          port :: port(),
-         map :: #{ {integer(), file:filename()} => any(),
-                   file:filename() => integer()}
+         watch_map :: #{
+           {watch_descriptor(), file:filename()} => any(),
+            directory()                          => watch_descriptor()
```
Comment (RAttab) [inline src/filewatch.erl:22]: Am I correct in assuming that this map spec indicates that there are two possible key types for this one map which both map to different value types? If so I'm curious why we're not using two maps, one for each key value type?

File: c_src/filewatch_inotify.c [hunk 13]
```diff
@@ -20,38 +29,46 @@ enum { MAX_COOLDOWNS = 12, MSG_LENGTH = 11 };
...
+static bool get_arguments(char *command, unsigned long *cooldown_ms)
+{
+    if (!command || !cooldown_ms) return false;
+
+    command = strchr(command, ' ');
+    if (!command) return false;
 
-static ErlDrvData start(ErlDrvPort port, char *UNUSED)
+    command += strspn(command, " ");
+    if (!*command) return false;
```
Comment (RAttab) [inline c_src/filewatch_inotify.c:53]: Might add a quick comment annotating what it is that you're skipping over here. Would avoid having to refer back to the erlang code to figure it out.

File: src/filewatch.erl (hunk 12 above)
Comment (rkallos) [reply src/filewatch.erl:22]: That's right. I stuck with using one map because I thought it'd be easier, and because I was certain there'd be no collisions between the different types of keys. I could split it out into two if you prefer.

**Comment by RAttab** [comment]
Few minor tweaks but otherwise it looks good. You may also want to rebase/squash the commits.

File: c_src/filewatch_inotify.c (hunk 13 above)
Comment (rkallos) [reply c_src/filewatch_inotify.c:53]: Done.

**Comment by rkallos** [comment]
Thanks for the review. I'll definitely rebase before merging
